File: modules/vescape-core/ios/sync/SyncAccepted.swift
import Foundation

/// The `200` body: what the server took, per table.
///
/// Validated exactly before any cursor moves. A missing table, an extra table, a non-integer count
/// or a count that differs from what was submitted is a protocol failure — the server applies a
/// batch whole, so anything else means the two sides disagree about what was stored, and advancing a
/// cursor on that disagreement is unrecoverable.
///
/// Parsed here rather than with `JSONSerialization` so the rule behaves identically on both
/// platforms.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sync/SyncAccepted.kt
enum SyncAccepted {
  /// Accepted counts by table, or nil when the body is not exactly the expected response.
  static func parse(_ body: String) -> [SyncTable: Int]? {
    var counts: [SyncTable: Int] = [:]
    var scanner = Scanner(body)
    guard scanner.expect("{"), scanner.expectKey("accepted"), scanner.expect("{") else { return nil }
    if scanner.peek() != "}" {
      while true {
        guard let name = scanner.string(), let table = SyncTable(rawValue: name) else { return nil }
        if counts[table] != nil || !scanner.expect(":") { return nil }
        guard let value = scanner.integer() else { return nil }
        counts[table] = value
        if scanner.expect(",") { continue }
        break
      }
    }
    guard scanner.expect("}"), scanner.expect("}"), scanner.atEnd() else { return nil }
    return counts.count == SyncTable.allCases.count ? counts : nil
  }

  /// True when the response accounts for exactly the rows submitted, table by table.
  static func matches(submitted: [SyncTable: Int], accepted: [SyncTable: Int]) -> Bool {
    SyncTable.allCases.allSatisfy { accepted[$0] == (submitted[$0] ?? 0) }
  }

  private struct Scanner {
    private let source: [Character]
    private var index = 0

    init(_ source: String) {
      self.source = Array(source)
    }

    mutating func atEnd() -> Bool {
      skipSpace()
      return index >= source.count
    }

    mutating func peek() -> Character? {
      skipSpace()
      return index < source.count ? source[index] : nil
    }

    mutating func expect(_ character: Character) -> Bool {
      guard peek() == character else { return false }
      index += 1
      return true
    }

    mutating func expectKey(_ name: String) -> Bool {
      string() == name && expect(":")
    }

    mutating func string() -> String? {
      guard expect("\"") else { return nil }
      var value = ""
      // Counts and table names carry no escapes; a body that needs them is not this response.
      while index < source.count, source[index] != "\"" {
        value.append(source[index])
        index += 1
      }
      guard index < source.count else { return nil }
      index += 1
      return value
    }

    mutating func integer() -> Int? {
      skipSpace()
      var digits = ""
      while index < source.count, source[index].isNumber {
        digits.append(source[index])
        index += 1
      }
      return digits.isEmpty ? nil : Int(digits)
    }

    private mutating func skipSpace() {
      while index < source.count, source[index].isWhitespace { index += 1 }
    }
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sync/SyncAccepted.kt
package expo.modules.vescapecore.sync

/**
 * The `200` body: what the server took, per table.
 *
 * Validated exactly before any cursor moves. A missing table, an extra table, a non-integer count or
 * a count that differs from what was submitted is a protocol failure — the server applies a batch
 * whole, so anything else means the two sides disagree about what was stored, and advancing a cursor
 * on that disagreement is unrecoverable.
 *
 * Parsed here rather than with the platform JSON so the rule runs in plain unit tests and behaves
 * identically on both platforms.
 *
 * @parity /modules/vescape-core/ios/sync/SyncAccepted.swift
 */
object SyncAccepted {
  /** Accepted counts by table, or null when the body is not exactly the expected response. */
  fun parse(body: String): Map<SyncTable, Int>? {
    val counts = LinkedHashMap<SyncTable, Int>()
    val scanner = Scanner(body)
    if (!scanner.expect('{') || !scanner.expectKey("accepted") || !scanner.expect('{')) return null
    if (scanner.peek() != '}') {
      while (true) {
        val name = scanner.string() ?: return null
        val table = SyncTable.entries.firstOrNull { it.wire == name } ?: return null
        if (counts.containsKey(table) || !scanner.expect(':')) return null
        counts[table] = scanner.integer() ?: return null
        if (scanner.expect(',')) continue
        break
      }
    }
    if (!scanner.expect('}') || !scanner.expect('}') || !scanner.atEnd()) return null
    return if (counts.size == SyncTable.entries.size) counts else null
  }

  /** True when the response accounts for exactly the rows submitted, table by table. */
  fun matches(submitted: Map<SyncTable, Int>, accepted: Map<SyncTable, Int>): Boolean =
    SyncTable.entries.all { accepted[it] == (submitted[it] ?: 0) }

  private class Scanner(private val source: String) {
    private var index = 0

    fun atEnd(): Boolean = skipSpace().let { index >= source.length }

    fun peek(): Char? = skipSpace().let { source.getOrNull(index) }

    fun expect(char: Char): Boolean {
      if (peek() != char) return false
      index += 1
      return true
    }

    fun expectKey(name: String): Boolean = string() == name && expect(':')

    fun string(): String? {
      if (!expect('"')) return null
      val end = source.indexOf('"', index)
      // Counts and table names carry no escapes; a body that needs them is not this response.
      if (end < 0) return null
      return source.substring(index, end).also { index = end + 1 }
    }

    fun integer(): Int? {
      skipSpace()
      val start = index
      while (index < source.length && source[index].isDigit()) index += 1
      return if (index == start) null else source.substring(start, index).toIntOrNull()
    }

    private fun skipSpace() {
      while (index < source.length && source[index].isWhitespace()) index += 1
    }
  }
}
